From df137b7d75ece656be23e3d2e43caf122a189f1b Mon Sep 17 00:00:00 2001
From: Thomas Frauendorfer <tf@miray.de>
Date: Thu, 13 Nov 2014 15:15:33 +0100
Subject: [PATCH 37/41] remove '*' from highlight

---
 grub-core/normal/menu_text.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/grub-core/normal/menu_text.c b/grub-core/normal/menu_text.c
index 77d3f67..2d0c1fd 100755
--- a/grub-core/normal/menu_text.c
+++ b/grub-core/normal/menu_text.c
@@ -264,7 +264,8 @@ print_entry (int y, int highlight, grub_menu_entry_t entry,
       unicode_title[i] = ' ';
 
   if (data->geo.num_entries > 1)
-    grub_putcode (highlight ? '*' : ' ', data->term);
+    grub_putcode (' ', data->term);
+    //grub_putcode (highlight ? '*' : ' ', data->term);
 
   grub_print_ucs4_menu (unicode_title,
 			unicode_title + len,
-- 
1.9.2.msysgit.0

